File: backend/src/main/java/com/portfolioapp/portfoliobackend/mapper/PhotoMapper.java
package com.portfolioapp.portfoliobackend.mapper;

import com.portfolioapp.portfoliobackend.dto.PhotoDto;
import com.portfolioapp.portfoliobackend.entity.Photo;
import com.portfolioapp.portfoliobackend.entity.UserProfile;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.factory.Mappers;

@Mapper(componentModel = "spring")
public interface PhotoMapper {
    PhotoMapper INSTANCE = Mappers.getMapper(PhotoMapper.class);

    @Mapping(target = "userProfile", ignore = true)
    @Mapping(target = "likedByUsers", ignore = true)
    Photo toEntity(PhotoDto dto);

    default PhotoDto toDto(Photo entity, UserProfile currentUser) {
        String fullName = (entity.getUserProfile() != null) ? entity.getUserProfile().getFullName() : null;
        String avatarUrl = (entity.getUserProfile() != null) ? entity.getUserProfile().getAvatarUrl() : null;

        return PhotoDto.builder()
                .id(entity.getId())
                .title(entity.getTitle())
                .description(entity.getDescription())
                .imageUrl(entity.getImageUrl())
                .author(entity.getAuthor())
                .authorFullName(fullName)          
                .authorAvatarUrl(avatarUrl)       
                .createdAt(entity.getCreatedAt())
                .updatedAt(entity.getUpdatedAt())
                .likesCount(entity.getLikesCount())
                .likedByCurrentUser(entity.isLikedBy(currentUser))
                .build();
    }

    default PhotoDto toDto(Photo entity) {
        return toDto(entity, null);
    }

    void updateEntityFromDto(PhotoDto dto, @MappingTarget Photo entity);
}

File: app/src/main/java/com/example/portfolioapp/data/local/mapper/PhotoMapper.kt
package com.example.portfolioapp.data.local.mapper

import android.net.Uri
import com.example.portfolioapp.entity.Photo
import com.example.portfolioapp.data.local.entity.PhotoEntity
import com.example.portfolioapp.network.PhotoDto

fun PhotoDto.toEntity(): PhotoEntity {
    return PhotoEntity(
        id = this.id.toInt(),
        title = this.title,
        description = this.description ?: "",
        imageUrl = this.imageUrl,
        likes = this.likesCount,
        isLiked = this.likedByCurrentUser,
        authorEmail = this.author,
        authorFullName = this.authorFullName ?: this.author.substringBefore("@", "Пользователь"),
        authorAvatarUrl = this.authorAvatarUrl
    )
}

fun PhotoEntity.toDomain(): Photo {
    return Photo(
        id = this.id,
        title = this.title,
        description = this.description,
        uri = Uri.parse(this.imageUrl),
        likes = this.likes,
        isLiked = this.isLiked,
        authorEmail = this.authorEmail,
        authorFullName = this.authorFullName,
        authorAvatarUri = this.authorAvatarUrl?.let { Uri.parse(it) },
        isLocal = false
    )
}

fun Photo.toEntity(): PhotoEntity {
    return PhotoEntity(
        id = this.id,
        title = this.title,
        description = this.description,
        imageUrl = this.uri?.toString() ?: "",
        likes = this.likes,
        isLiked = this.isLiked,
        authorEmail = this.authorEmail,
        authorFullName = this.authorFullName,
        authorAvatarUrl = this.authorAvatarUri?.toString()
    )
}